packages/openactive-integration-tests/test/helpers/flow-stages/opportunity-feed-update.js

```diff
@@ -1,162 +1,319 @@
 const sharedValidationTests = require('../../shared-behaviours/validation');
 const { isResponse20x } = require('../chakram-response-utils');
+const { FlowStage } = require('./flow-stage');
 
 /**
  * @typedef {import('chakram').ChakramResponse} ChakramResponse
  * @typedef {import('../../types/OpportunityCriteria').OpportunityCriteria} OpportunityCriteria
  * @typedef {import('../logger').BaseLoggerType} BaseLoggerType
  * @typedef {import('../request-helper').RequestHelperType} RequestHelperType
+ * @typedef {import('./flow-stage').FlowStageOutput} FlowStageOutput
  */
 
-const { HARVEST_START_TIME } = global;
+/**
+ * @typedef {Required<Pick<FlowStageOutput, 'testInterfaceOpportunities'>>} ListenerInput
+ * @typedef {Required<Pick<FlowStageOutput, 'getOpportunitiesFromOpportunityFeedPromise'>>} ListenerOutput
+ *
+ * @typedef {ListenerOutput} CollectorInput
+ * @typedef {Required<Pick<FlowStageOutput, 'opportunityFeedExtractResponses'>>} CollectorOutput
+ */
+
+/**
+ * @typedef {'wait-for-one' | 'wait-for-all'} WaitMode
+ */
+
+/**
+ * @param {object} args
+ * @param {ChakramResponse[]} args.testInterfaceOpportunities
+ * @param {RequestHelperType} args.requestHelper
+ * @param {boolean} [args.useCacheIfAvailable]
+ */
+async function fetchOpportunityFeedExtractResponses({ testInterfaceOpportunities, requestHelper, useCacheIfAvailable }) {
+  // TODO Note that waitMode is not yet used in this function. It should be. But, since
+  // waitMode=wait-for-one only supports one opportunity, wait-for-one & wait-for-all
+  // are exactly equivalent, until waitMode=wait-for-one can support multiple opportunities.
+  /**
+   * Note that the reponses are stored wrapped in promises. This is because
+   * opportunities are created/fetched concurrently.
+   * If the responses were awaited before storing them in this cache, then
+   * the same request may be made over and over.
+   *
+   * @type {Map<string, Promise<ChakramResponse>>}
+   */
+  const reusableMatchPromises = new Map();
+
+  /**
+   * Full opportunity data for each opportunity fetched by fetchOpportunities() - one for each criteria.
+   *
+   * Each is an RPDE item.
+   */
+  const opportunityFeedExtractResponses = await Promise.all(testInterfaceOpportunities.map(async (testInterfaceOpportunity, i) => {
+    // Only attempt getMatch if test interface response was successful
+    if (isResponse20x(testInterfaceOpportunity) && testInterfaceOpportunity.body['@id']) {
+      const opportunityId = testInterfaceOpportunity.body['@id'];
+      // If a match for this @id is already being requested, just reuse the same response
+      if (reusableMatchPromises.has(opportunityId)) {
+        return await reusableMatchPromises.get(opportunityId);
+      }
+
+      const matchPromise = requestHelper.getMatch(opportunityId, i, useCacheIfAvailable);
+      reusableMatchPromises.set(opportunityId, matchPromise);
+      return await matchPromise;
+    }
+    return null;
+  }));
```

Comment:
This logic is unchanged. I just moved it from a function called `OpportunityFeedUpdateFlowStage.run` to this function